### ISSUE TITLE
The pre-cache callbacks loop should be parallelized

### ISSUE THREAD

**Issue by jmarca** [google/or-tools#2112]
I've been testing out some slow code.  To make a long story short, I've been setting  `model_parameters.max_callback_cache_size` greater than number of nodes so as to always trigger caching.  

Anyway, in doing that I've found that this loop is super slow:

https://github.com/google/or-tools/blob/c59101fc9e7b4dda50429cad5a291f13b580a2c9/ortools/constraint_solver/routing.cc#L826
Copying here: 
https://github.com/google/or-tools/blob/c59101fc9e7b4dda50429cad5a291f13b580a2c9/ortools/constraint_solver/routing.cc#L823-L830

In my test case, I have 5500 vehicles and 1041 nodes, so that is a large matrix, and running that loop takes 15 minutes. 

**But** it uses just one core, and I've got 7 others sitting idle.  It seems to me this simple loop could be parallelized to get a significant speedup, but I know nothing about the intricacies of doing so in C++, and searching quickly though the code base doing so might be a headache.  But as a user I know there is parallel support in the cp_sat side of things, so there must be a reasonable set of conventions somewhere.

Comparing run times, the precache version is faster.

With precaching,  after the 15 minutes to call `routing.RegisterTransitCallback(time_callback)`, the solver finds the 
 All Unperformed Solution (1036000000000) at time = 11794 ms (11 seconds), and stabilizes on its final solution at  time = 124455 ms (about 2 minutes).

Without precaching (using defaults of cache=1000), the solver finds the  All Unperformed Solution (1036000000000) at time = 439661 ms (7.3 minutes) and ~~has not yet stabilized on the best solution before hitting 12.6 minutes (I set total solver time to 20 minutes)~~ (update with a 45min run) took a further 33.7 minutes to stabilize on the best solution. 

So with caching:    15 minutes   + 11 seconds + 2 minutes   = about 17 minutes 
without caching:  no cache time + 7 minutes  + 33 minutes  = about 40 minutes to reach same solution

If the 15 minutes can be parallelized to 8 threads, I'd expect around 3 minutes for that step (allowing for marshaling the results), for a total run time of about 4 minutes.  

So definitely worth it, in my opinion.

**Comment by jmarca** [comment]
Googling "how to program parallel loops in C++" is a terrible idea, but...is there a reason why the parallel versions of the std template library for each wouldn't work here?

Something like 
```cpp
std::vector<int64> cache(size * size, 0);
// is there a way to initalize the cache to equal the index number, so cache[0] === 0 and cache[50] === 50, etc?
// if so do that, then 
std::for_each(std::execution::par,
                      cache.begin(), 
                      cache.end(), 
                      f);
```

where f is some function that take an element of the vector and modifies it in place.
If the element of the cache vector is initialize to the index, then f can figure out i and j from the value of the passed in vector element (v) doing something like:

```c++
f(v){
  // assuming value of v is initialized to the index
  j = v % size;
  // is there a div operator in C++?
  i = (v - j) / size; // or integer divide?  
  v = callback(i,j)
  // bob's your uncle
```

But I haven't done C++ and STL stuff since 1992 or so.  Things have changed.

**Comment by Mizux** [comment]
unfortunately ExecutionPolicy comes in C++17 i.e. STL just discover the multi-threading world ! Unfortunately we still try to target C++11 which is widely available.

On the other side we could use pthread which is here for long time but a little bit a pain to use.

Last proposal would be to use OpenMP pragma which is less intrusive but don't know how well is supported on all paltforms.

note: I was thinking of something like
```cpp
std::vector<int> c(SIZE * SIZE, 0);
std::iota(c.begin(), c.end(), 0); // fill vector with 1,2,3,4,5,6 but not sure is paralellized

std::transform(
  std::execution::par,
  c.cbegin(), c.cend(),
  c.begin(),
  [](int i) -> int { return callback(i / SIZE, i % SIZE); }); // can be expensive to compute i and j
```

**Comment by jmarca** [comment]
Working from the above suggestion, I got the following to compile:

```c++
int RoutingModel::RegisterTransitCallback(TransitCallback2 callback) {
  if (cache_callbacks_) {
    const int size = Size() + vehicles();
    std::vector<int> cache(size * size, 0);
    std::iota(cache.begin(), cache.end(), 0); // fill vector with 1,2,3,4,5,6 but not sure is paralellized

    std::transform(
                   std::execution::par,
                   cache.cbegin(), cache.cend(),
                   cache.begin(),
                   [callback, size](int i) -> int { return callback(i / size, i % size); });
    transit_evaluators_.push_back(
        [cache, size](int64 i, int64 j) { return cache[i * size + j]; });
  } else {
    transit_evaluators_.push_back(std::move(callback));
```

Compilation succeeded by switching OR Tools to C++17, adding the intel Tbb library (libtbb) to get the std::execution::par to work, and fixing the compilation of all dependencies to also use C++17 (abseil was resistant).  All of the C++ and Python tests run successfully.

But when I run my code that triggers this precache algorithm, I get an unceremonious segfault.

I added a print statement in my callback, and I'm getting sane values for from_index and to_index.  The crash is definitely happening on the C++ side, after anywhere between 3 and 7 calls to the callback are made.

So two lessons learned:  apparently none of the existing test programs use `cache_callbacks_`, and something in the above is broken.  

Which means I have to figure out another way to speed up C++ talking to the python callback function.  Are there some SWIG techniques I can leverage?

**Comment by jmarca** [comment]
Closing for now until I come up with a parallel loop that works

**Comment by jmarca** [comment]
One more note to this.  I kept trying different things on this, and I've come to the conclusion that SWIG does not play well with `std::execution::par` or `std::execution::par_unseq`.

I think this because when I commented out that line in the C++, it ran fine...so the bug isn't the algorithm or in the callback, but rather in what is going on when it is split up to run in parallel.

That and the fact that there is zero documentation that I can find about using `std::execution::par` with SWIG callbacks.  If no one has written anything about it, the most likely no one thought to handle that case in SWIG at all.

**Comment by Mizux** [comment]
my bet on a Python GIL issue than a SWIG one itself

**Comment by jmarca** [comment]
Another possibility is something like that suggested in #321.  Passing the matrix of data directly to the solver and let the solver generate a simple lookup callback.  When/If somebody tackles that, perhaps similar approach could be used here.

**Comment by jmarca** [comment]
This issue is still tripping me up.  I've been trying to work around it because often it is just foolish to try to solve a very large problem, but for example, today I was trying to fit 4 new destinations in a large problem that I had already solved by decomposition.  So most of the nodes are locked to vehicles, and there isn't much work to do, but I had to wait 600+ seconds to load the data into C++.

Because the parallel thing isn't going to work, I'm not going to reopen this issue, but how about passing data to C++, rather than a callback, and let C++ create the callback itself (as is done when precaching the callback).  I will try to work on that, but my C++ skills are still rusted shut.
